crates/zome_types/src/lib.rs

```diff
@@ -73,3 +75,121 @@ pub trait CallbackResult {
+/// Being able to wrap fixed size arrays is important e.g. for crypto safety etc. so this is a
+/// simple way to implement serialization so that we can send these types between the host/guest.
+macro_rules! fixed_array_serialization {
+    ($t:ty, $len:expr) => {
+        impl serde::ser::Serialize for $t {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::ser::Serializer,
+            {
+                serializer.serialize_bytes(&self.0)
+            }
+        }
+
+        impl<'de> serde::de::Deserialize<'de> for $t {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: serde::de::Deserializer<'de>,
+            {
+                use serde::de::Error;
+                let bytes: &[u8] = serde::de::Deserialize::deserialize(deserializer)?;
+                if bytes.len() != $len {
+                    let exp_msg = format!("expected {} bytes got: {} bytes", $len, bytes.len());
+                    return Err(D::Error::invalid_value(
+                        serde::de::Unexpected::Bytes(bytes),
+                        &exp_msg.as_str(),
+                    ));
+                }
+                let mut inner: [u8; $len] = [0; $len];
+                inner.clone_from_slice(bytes);
+                Ok(Self(inner))
+            }
+        }
+    };
+}
+
+#[macro_export]
+/// Cryptographic secrets are fiddly at the best of times.
+///
+/// In wasm it is somewhat impossible to have true secrets because wasm memory is not secure.
+///
+///  - The host can always read wasm memory so any vulnerability in the host compromises the guest.
+///  - The host/rust generally doesn't guarantee to immediately wipe/zero out freed memory, either
+///    when a zome call is running or after a wasm instance is thrown away.
+///
+/// Most of the time we should just try to minimise the interaction between wasm and secret data.
+///
+/// For example, lair keeps all our private keys internal and we can only send it signing requests
+/// associated with public keys.
+///
+/// In other contexts it is more difficult, such as when generating secrets from raw cryptographic
+/// random bytes and sending them to peers directly.
+///
+/// The best we can do here is try to protect ourselves against third parties across the network.
+/// e.g. We don't want other machines to simply `remote_call` a successful timing attack.
+///
+/// MITM attacks are mitigated by the networking implementation itself.
+///
+/// @todo given how impossible it is for wasm to protect its memory from the host, it would make
+/// more sense to:
+///
+///  - use key exchange protocols like libsodium kx https://libsodium.gitbook.io/doc/key_exchange.
+///  - keep secrets inside lair with all algorithms behind an API, wasm only has access to opaque
+///    references to the secret data.
```

Comment:
Either / both of these are totally valid strategies to base our API on - we don't even need lair to be implemented yet, we can keep the private keys unprotected in holochain for the moment, but write the WASM api properly to interact with handles or key exchange mechanisms so app devs don't go down an insecure path with their code, right?